### ISSUE TITLE
Feature request: add parametric ELU (PELU) activation function

### ISSUE THREAD

**Issue by carlthome** [tensorflow/tensorflow#7712]
The exponential linear unit (ELU) is already in TensorFlow as `tf.nn.elu` which is great. The new parametric version (called PELU) shows very promising experimental results so I wonder if it could be added in to TensorFlow too in order to encourage more widespread experimentation with it by the deep learning community. One problem with it though is that it's stateful (e.g. `tf.Variable`), meaning it's not clear to me where in TensorFlow it fits in.

Here's an implementation of the PELU that I've been toying around with:
```python
def pelu(x):
  """Parametric Exponential Linear Unit (https://arxiv.org/abs/1605.09332v1)."""
  with tf.variable_scope(x.op.name + '_activation', initializer=tf.constant_initializer(1.0)):
    shape = x.get_shape().as_list()[1:]
    alpha = tf.get_variable('alpha', shape)
    beta = tf.get_variable('beta', shape)
    positive = tf.nn.relu(x) * alpha / (beta + 1e-9)
    negative = alpha * (tf.exp((-tf.nn.relu(-x)) / (beta + 1e-9)) - 1)
    return negative + positive
```

**Comment by poxvoculi** [comment]
@fchollet: Is this a welcome contribution?

**Comment by fchollet** [comment]
Anything that has a state (e.g. this function) should not be added as a function `tf.nn` (which should only contain pure functions), but rather as a layer (`tf.layers`). This is precisely what layers are meant to cover. This is a `PELU` layer, not an activation function.

However I would be wary of adding new core layers for each and every paper out there. At best this would be a candidate for contrib.

**Comment by carlthome** [comment]
> Anything that has a state (e.g. this function) should not be added as a function in tf.nn (which should only contain pure functions), but rather as a layer (tf.layers). This is precisely what layers are meant to cover. This is a PELU layer, not an activation function.

Yeah, figured `tf.nn` should be without side effects. Good to hear! :+1: 

> However I would be wary of adding new core layers for each and every paper out there. At best this would be a candidate for contrib.

Which contrib package would be suited for new activation functions? `tf.contrib.layers` is still out-of-sync with `tf.layers` (both have similar purpose but different origins and APIs) so I expect it might be deprecated or heavily refactored soon? It would be nice if `tf.contrib.layers` were an extension for `tf.layers`, but I suppose it's hard to juggle the backwards compatibility.

**Comment by carlthome** [comment]
A little warning: the PELU implementation above might cause NaN loss when it becomes too steep (inf gradient). :koala:

**Comment by rmkemker** [comment]
@carlthome, where are you implementing the non-negativity constraint for a and b?  Just in the loss function?

**Comment by carlthome** [comment]
Derp. Good catch, @rmkemker. That's why I've been sometimes getting NaNs...

I guess just penalizing the loss could work
```python
loss += tf.minimum(tf.reduce_min(alpha), 0)
loss += tf.minimum(tf.reduce_min(beta), 0)
```
but how about just wrapping the parameters in `tf.abs` and hoping it doesn't matter that two points on the loss surface map to the same ELU shape?

**Comment by rmkemker** [comment]
That is exactly what I did, but I am still getting the NaN values after it trains for a while.  I even tried weighting the penalties a bit higher with the same result.  My code trains till completion with ReLU activation, so it must be the PELU somehow.

**Comment by carlthome** [comment]
[SELU](https://arxiv.org/abs/1706.02515) feels like a big improvement over ReLU (without needing extra parameters) so I don't believe PELU is relevant anymore. As far as I'm concerned this issue could be closed.

**Comment by rmkemker** [comment]
I will give it a shot.  As a FYI, I was able to stabilize training for PELU in some instances by lowering the learning rate.  It is still pretty finicky.  Hopefully, SELU is more stable during training!

**Comment by kacper1095** [comment]
Is it a good implementation of the PELU? In the paper there are lines like: 
```
Unlike Maxout, our PELU adds only 2L parameters, where
L is the number of layers, which makes our activation as
computationally demanding as the original ELU function
``` 
and
 ```
It is interesting to note that PELU only adds 112 additional
parameters, a negligible increase of 0.006% over the total
number of parameters.
``` 
(regarding to ResNet-112). Still, your implementation introduces one parameter for each neuron, which is much more then the number mentioned in the paper. I suggest following change from:
```python
    shape = x.get_shape().as_list()[1:]
    alpha = tf.get_variable('alpha', shape)
    beta = tf.get_variable('beta', shape)
```
to
```python
    alpha = tf.get_variable('alpha', 1)
    beta = tf.get_variable('beta', 1)
```
Please, correct me if I didn't understand the paper correctly.

**Comment by carlthome** [comment]
Thanks for correcting this! 😃

**Comment by SSaishruthi** [comment]
Hello,
Is anyone working on this?
TIA

**Comment by dynamicwebpaige** [comment]
Thank you for suggesting this feature request! Though parametric ELU would not necessarily fit in `tensorflow/tensorflow`, I'm sure it would be a welcome contribution to activations in [Addons](https://github.com/tensorflow/addons/tree/master/tensorflow_addons/activations).

@seanpmorgan for awareness; am going to transfer this issue to `tensorflow/addons`.
